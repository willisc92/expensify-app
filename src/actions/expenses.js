import uuid from 'uuid';
import database from '../firebase/firebase'

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// With firebase:
// component calls action generator
// action generator returns function
// component dispatches function
// function runs (has ability to dispatch other actions)


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid; 
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = {description, note, amount, createdAt};
        
        return database.ref(`users/${uid}/Expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense 
            }));
        });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id 
});

export const startRemoveExpense = ({id} = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/Expenses/${id}`).remove().then((ref) => {
            dispatch(removeExpense({id}));
        });
    };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/Expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid; 
        return database.ref(`users/${uid}/Expenses`)
            .once('value')
            .then((snapshot) => {
                const expenses = [];

                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch(setExpenses(expenses));
            });   
    };
};