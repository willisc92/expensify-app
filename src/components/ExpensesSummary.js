import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import getExpenseTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({expenseCount, expensesTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal/100).format('$0,0.00');

    return(
        <div>
            <h1>Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}</h1>
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpenseTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);