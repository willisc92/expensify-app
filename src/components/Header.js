import React from 'react';
import {NavLink} from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <div><NavLink to="/" activeClassName="is-active" exact={true}>Home</NavLink></div>
        <div><NavLink to="/create" activeClassName="is-active" exact={true}>Create Expense</NavLink></div>
    </header>
)

export default Header;