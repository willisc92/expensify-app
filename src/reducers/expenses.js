// Expenses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense]; // array spread operator syntax- equivalent to state.concat(action.expense).
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates   // object spread operator - copies/overrides key-values from object(s) into the new object
                        // requires "babel-plugin-transform-object-rest-spread" plugin
                    }
                } else {
                    return expense;
                }
            });
        default:
            return state;
    }
};

