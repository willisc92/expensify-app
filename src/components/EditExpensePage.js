import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
    return (
        <div>
            <ExpenseForm 
                expense = {props.expense}
                onSubmit = {(expense) => {
                    props.dispatch(editExpense(props.expense.id, expense));
                    props.history.push('/');
                }}
            />
            <button onClick = {() => {
                props.dispatch(removeExpense({id: props.expense.id})); 
                props.history.push('/');
            }}>Remove</button>
        </div>
    );
};

const mapStatetoProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

export default connect(mapStatetoProps)(EditExpensePage);

// React router passes down components passed down in a route, including:
// history
// location - for use with hash values (example: localhost:8080/edit#contact-us) or search (example: localhost:8080/edit?query-rent&sort=date)
// match (for dynamic pages)