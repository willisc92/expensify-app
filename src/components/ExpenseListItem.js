import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({id, dispatch, description, amount, createdAt}) => (
    <div>
        <h3>{description}</h3>
        <p>Amount: ${amount/100} - Created At: {createdAt}</p>
        <button onClick = {() => {
           dispatch(removeExpense({id})); 
        }}>Remove</button>
    </div>
);

export default connect()(ExpenseListItem);