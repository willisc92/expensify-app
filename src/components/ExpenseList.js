import React from "react";
import { connect } from "react-redux"; // Needed wherever components need to connect to the store to perform actions or read from the store.
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>

        <div className="list-body">
            {props.expenses.length === 0 ? (
                <div className="list-item list-item--message">
                    <span>No expenses</span>
                </div>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense} />;
                })
            )}
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
}; // Function that maps the state provided by the store to props to be used in the component.

export default connect(mapStateToProps)(ExpenseList); // connect returns the function that returns the HOC.
// argument inside connect should return an object defining what subset from the store that should be accessed.
