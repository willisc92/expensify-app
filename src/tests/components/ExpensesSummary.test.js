import { ExpensesSummary } from "../../components/ExpensesSummary";
import React from "react";
import expenses from "../fixtures/expenses";
import { shallow } from "enzyme";
import getExpenseTotal from "../../selectors/expenses-total";

test("Should test Expense Summary for one expenses", () => {
    const singleExpense = [expenses[0]];
    const expensesTotal = getExpenseTotal(singleExpense);
    const wrapper = shallow(<ExpensesSummary expenseCount={singleExpense.length} expensesTotal={expensesTotal} />);
    expect(wrapper).toMatchSnapshot();
});

test("Should test Expense Summary for multiple expenses", () => {
    const expensesTotal = getExpenseTotal(expenses);
    const wrapper = shallow(<ExpensesSummary expenseCount={expenses.length} expensesTotal={expensesTotal} />);
    expect(wrapper).toMatchSnapshot();
});
