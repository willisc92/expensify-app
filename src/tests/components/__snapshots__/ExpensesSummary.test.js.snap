// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should test Expense Summary for multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
         
      </span>
       
      expenses
       totalling 
      <span>
        $1,141.95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should test Expense Summary for one expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
         
      </span>
       
      expense
       totalling 
      <span>
        $1.95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
